### PR TITLE
66 scotland features

### LINKED ISSUE #46: Add script to download INSPIRE files
Add script to repo to automatically download INSPIRE land registry files for England and Wales from gov site: https://use-land-property-data.service.gov.uk/datasets/inspire/download

### LINKED ISSUE #66: Add features for Scotland
Features to add for Scotland:

- Land registry parcels
- Listed buildings
- Building conservation areas
- Property density

### PR DESCRIPTION
**Pull request by crispy-wonton**
Fixes #66 and #46
---

# Description

Add features (reweighting features added in a separate branch) for Scotland:
- Garden size
- Listed building status
- Property in World Heritage Site
- Property density of Data Zone
NB: I need to add the new datasets to the README. This will be added in a separate PR.

Changes by feature:

**All features**
- [UPDATED] `pipeline/run_scripts/run_add_features.py` - update script to add new Scottish features
- [UPDATED] `getters/get_datasets.py` - update listed buildings getter to load Scotland data & add getters for Scotland data zone boundaries and number of dwellings
- [UPDATED] `getters/base_getters.py` - minor changes to existing functions
- [UPDATED] `config/base.yaml` - new data sources for Scotland added

**Garden size estimates**
- [NEW] `pipeline/run_scripts/run_get_inspire_file_bounds.py` - script to generate bounding polygons of each INSPIRE land registry file and save to S3.
- [NEW] `pipeline/run_scripts/run_stream_inspire_files.py` - script to stream INSPIRE land registry files for Scotland and/or England & Wales from their respective government websites and into S3.
- [UPDATED] `pipeline/prepare_features/building_footprint.py` - update existing function to convert footprints to BNG with a try/except block. This was added because when running the pipeline for Scotland, there was at least one polygon that appeared to be corrupted and threw a shapely error during processing.
- [UPDATED] `pipeline/prepare_features/land_extent.py` - add function to get file bounds for Scotland land registry files; update variable names in existing function for readability; and update function to transform land registry files to transform Scottish as well as England and Wales data
- [UPDATED] `pipeline/run_scripts/run_calculate_garden_size.py` - update script to calculate garden size for any of England & Wales/Scotland specified in user input.

**Listed buildings**
- [NEW] `notebooks/20240926_explore_scotland_listed_buildings.py` - calculate threshold for `sjoin_nearest` for joining UPRNs with listed buildings
- [UPDATED] `pipeline/prepare_features/listed_buildings.py` - add function to identify listed buildings in EPC for any of England/Scotland/Wales; and add function to conduct `sjoin` or `sjoin_nearest` of UPRNs with listed building geometries depending on geometry type identified.
- [UPDATED] `pipeline/suitability/calculate_suitability.py` - remove redundant code.

**World Heritage Site**
- [UPDATED] `pipeline/prepare_features/protected_areas.py` - renamed from `conservation_areas.py` with functions to load and transform Scottish World Heritage Site data added 

**Property density**
- `pipeline/prepare_features/property_density.py` - add functions to transform Data Zone area and dwellings data and to generate property density dataframe for Scotland
- [UPDATED] `getters/schemas.py` - add schema for NRS dwellings data

**Other changes**
- [UPDATED] `pipeline/prepare_features/output_areas.py` - minor update to existing function to update input param
- [UPDATED] `requirements.txt` - add new requirements and update `numpy` version to avoid clashes with `pandas`

# Instructions for Reviewer

Please could you pay special attention to:
From draft PR and first review:
- the join logic for listed buildings, conservation areas, and world heritage sites
- the notebook analysis to check the threshold has been calculated and selected appropriately
- review the nearest neighbour join logic

From full PR please could you pay special attention to:
- check appropriate datasets used and that they have been processed correctly
- garden size estimate pipeline for Scotland to ensure it's been calculated correctly as the results are pretty large average garden sizes (median is over 1000m2).
- appropriate dwelling total has been used to calculate property density. Should we exclude any other dwelling categories from the total count?
- Double check all required features have been added

# Running the scripts
Instructions to run each script can be found at the top of each script. Here are the full commands for running for the 2023 Q4 EPC data:
Add features:
`python -i asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py --epc_path s3://asf-heat-pump-suitability/outputs/2023Q4/20240824_2023_Q4_EPC_weighted.parquet -y 2023 -q 4`

Stream INSPIRE files for Scotland
`python -i asf_heat_pump_suitability/pipeline/run_scripts/run_stream_inspire_files.py -n s`

Get INSPIRE file bounds for Scotland
`python -i asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py -y 2023 -q 4 -n s`

Calculate garden size for Scotland
`python -i asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_garden_size.py --epc_path s3://asf-heat-pump-suitability/outputs/2023Q4/20240824_2023_Q4_EPC_weighted.parquet -y 2023 -q 4 -n s`

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py [hunk 1]
```diff
@@ -47,18 +69,57 @@ def sjoin_df_epc_with_listed_buildings(
...
+    epc_gdf = lat_lon.generate_gdf_uprn_coords(df=epc_df, usecols=["UPRN"])
+    if any(
+        [
+            expr in listed_buildings_gdf.geom_type.unique()
+            for expr in ["Point", "MultiPoint"]
+        ]
+    ):
+        df = epc_gdf.sjoin_nearest(
+            listed_buildings_gdf, how="inner", max_distance=distance
+        )[["UPRN", "listed_building"]].drop_duplicates(subset="UPRN")
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py:109]: I think this looks right.

Thinking aloud: So you find all the points or geometries of listed properties within a certain distance of each UPRN, then drop duplicates (since you don't care about which listed property they match to, just that they match to at least one).

File: asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py [hunk 2]
```diff
@@ -47,18 +69,57 @@ def sjoin_df_epc_with_listed_buildings(
...
+    elif any(
+        [
+            expr in listed_buildings_gdf.geom_type.unique()
+            for expr in ["Polygon", "MultiPolygon"]
+        ]
+    ):
+        df = epc_gdf.sjoin(listed_buildings_gdf, how="inner", predicate="intersects")[
+            ["UPRN", "listed_building"]
+        ].drop_duplicates(subset="UPRN")
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py:118]: I assume `sjoin` can be used for both points and polygons, so why use `sjoin_nearest` at all?

File: asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py [hunk 3]
```diff
@@ -85,10 +85,11 @@ def parse_arguments() -> argparse.Namespace:
     logging.info("Adding lat/lon data to EPC")
     uprn_latlon_df = lat_lon.transform_df_osopen_uprn_latlon()
     epc_df = epc_df.join(uprn_latlon_df, how="left", on="UPRN")
+    epc_gdf = lat_lon.generate_gdf_uprn_coords(epc_df, usecols=["UPRN", "lad_code"])
 
     # Replace `lad_code` from postcode with `lad_code` from geospatial join and postcode
     logging.info("Adding LAD code with geospatial join")
-    uprn_lad_df = output_areas.sjoin_df_uprn_lad_code(epc_df)
+    uprn_lad_df = output_areas.sjoin_df_uprn_lad_code(epc_gdf)
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:93]: I've checked through all the joins in this script + the functions that are referenced. The joins make sense to me! At first I was wondering why you'd not just left joined everything to `epc_gdf` (rather than inner join to `epc_gdf` and then left join of result to `epc_df`), but I think the way youve done it is more clear and efficient.

**Review by lizgzil — COMMENTED** [review]
Hey @crispy-wonton - this looks great :) ! The joins all made sense to me, I just had a few comments.

I wonder whether the threshold of max_distance = 5 might be too harsh - but I can't remember your exact reasoning for it? I did a very brief analysis at the end of your notebook - and increasing it to 10 would allow many more of the correct matches to be outputted as a match (recall from 0.77 to 0.9), without effecting the precision too much (1 to 0.96). What do you think?

Screenshot 2024-10-11 at 15 48 04

File: asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py (hunk 3 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:93]: I think (?) it has to be an `sjoin_nearest` because with `sjoin`, these are the predicates for joining:
https://geopandas.org/en/stable/docs/user_guide/mergingdata.html#binary-predicate-joins
- intersects
- contains
- within
- touches
- crosses
- overlaps
So points that aren't exactly touching/ on top of each other will not get joined together. This was the case with the Wales listed buildings sjoin that was in the code before. It does join some of the buildings, but you get much more with the `sjoin_nearest` methodology. Does this make sense?

Comment (lizgzil) [reply asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:93]: yeh makes sense! Thanks for a explanation :)

File: asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py (hunk 1 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py:109]: Yes that's correct. So this should return a df with UPRNs and then a boolean flag for being in a listed building. Because it's an inner join, this will only return UPRNs which are in listed buildings.

File: asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py (hunk 2 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py:118]: Somehow I ended up responding to this question in the wrong comment reply. Copying my answer here for future ref:

I think (?) it has to be an `sjoin_nearest` because with `sjoin`, these are the predicates for joining:
https://geopandas.org/en/stable/docs/user_guide/mergingdata.html#binary-predicate-joins

intersects
contains
within
touches
crosses
overlaps
So points that aren't exactly touching/ on top of each other will not get joined together. This was the case with the Wales listed buildings `sjoin` that was in the code before. It does join some of the buildings, but you get much more with the `sjoin_nearest` methodology. Does this make sense?

**Comment by crispy-wonton** [comment]
> Hey @crispy-wonton - this looks great :) ! The joins all made sense to me, I just had a few comments.
> 
> I wonder whether the threshold of max_distance = 5 might be too harsh - but I can't remember your exact reasoning for it? I did a very brief analysis at the end of your notebook - and increasing it to 10 would allow many more of the correct matches to be outputted as a match (recall from 0.77 to 0.9), without effecting the precision too much (1 to 0.96). What do you think?

@lizgzil thanks for adding this analysis. I had set it to 5m because it was the 75th percentile of distance for true matches. I had thought we could increase to 11m to capture more listed buildings, which was the upper bound of the box plot. We discussed these options in the sprint review meeting and determined to stick with 5m although I'm not sure there was much additional reasoning beyond that, so I can update it to use 10m.

File: asf_heat_pump_suitability/pipeline/suitability/calculate_suitability.py [hunk 4]
```diff
@@ -190,14 +190,6 @@ def get_enhanced_epc(path) -> pl.DataFrame:
...
         ~pl.col("UPRN").str.contains("dummy"), pl.col("COUNTRY") != "Scotland"
     )
-
-    df = df.with_columns(
-        pl.when(pl.col("listed_building_grade").is_null())
-        .then(False)
-        .otherwise(True)
-        .alias("listed_building"),
-    )
-
```
Comment (crispy-wonton) [inline asf_heat_pump_suitability/pipeline/suitability/calculate_suitability.py:200]: This conversion is no longer needed because `pipeline/prepare_features/listed_buildings.py` functions have been updated to output a flag with `True`/`False` already for listed building status.

File: asf_heat_pump_suitability/pipeline/run_scripts/run_stream_inspire_files.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        ros_url = config["data_source"]["S_ros_inspire_url"]
+
+        page = requests.get(ros_url)
+        soup = BeautifulSoup(page.content, "html.parser")
+
+        # Generate download URLs for each Scottish registration county
+        pattern = re.compile(f'https:.+?(?=")')
+        url_prefix = pattern.search(soup.find("script", string=pattern).text).group(0)
+        ids = {v.contents[0]: v["value"] for v in soup.find_all("option") if v["value"]}
+        download_urls = {area: url_prefix + url for area, url in ids.items()}
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/run_scripts/run_stream_inspire_files.py:63]: nice! I didn't know about `soup.find_all("option")` - very handy!

File: asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        s_gdf = land_extent.generate_gdf_file_bounds_s(
+            config["data_source"]["S_inspire_land_extent_dir"]
+        )
+        save_as = f"s3://asf-heat-pump-suitability/outputs/{year}Q{q}/inspire_file_bounds_S.geojson"
+        s_gdf.to_file(save_as)
+
+    if args.nations == "ews":
+        gdf = pd.concat([ew_gdf, s_gdf]).reset_index()
+        save_as = f"s3://asf-heat-pump-suitability/outputs/{year}Q{q}/inspire_file_bounds_EWS.geojson"
+        s_gdf.to_file(save_as)
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py]: is the idea that by saving them in the EPC year and quarter folder, then we know exactly which INSPIRE files went into that particular version of the HP suitability data? (that would make sense, just wondering because the EPC data has nothing to do with these INSPIRE scripts)

File: asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        help="Nations to get INSPIRE land registry file bounds for. Of England and Wales (ew); Scotland (s); or all (ews).",
+        type=str,
+        choices=["ew", "s", "all"],
+        required=True,
+    )
+
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py:67]: I can only see `inspire_file_bounds_S.geojson` in the s3 bucket. Will this break anything downstream to have not have the data for the england and wales in this location too?

File: asf_heat_pump_suitability/pipeline/prepare_features/land_extent.py [hunk 8]
```diff
@@ -50,57 +78,52 @@ def _standardise_list_council_names(name_series: pd.Series) -> list:
...
 
     Returns:
         gpd.GeoDataFrame: land extent (INSPIRE) files and their bounding polygons in British National Grid CRS
     """
-    council_bounds = transform_gdf_council_bounds(ladnm_col, use_cols)
-    land_extent_file_names = base_getters.list_files_s3_location(land_extent_location)
+    bounds_gdf = transform_gdf_council_bounds(ladnm_col, use_cols)
+    files = base_getters.list_obj_s3_location(path)
 
     matches = _match_list_file_to_name(
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/prepare_features/land_extent.py:117]: I had a look at the outputs of this function, and times when it didn't match:
```
[(a,b) for a,b in zip(land_extent_files,matches) if not b]
[('asf-heat-pump-suitability/source_data/inspire_gml/', None),
('asf-heat-pump-suitability/source_data/inspire_gml/Borough_Council_of_Kings_Lynn_and_West_Norfolk_Land_Registry_Cadastral_Parcels.gml', None),
('asf-heat-pump-suitability/source_data/inspire_gml/Durham_County_Council_Land_Registry_Cadastral_Parcels.gml', None), ('asf-heat-pump-suitability/source_data/inspire_gml/Hull_City_Council_Land_Registry_Cadastral_Parcels.gml', None),
('asf-heat-pump-suitability/source_data/inspire_gml/Newcastle_City_Council_Land_Registry_Cadastral_Parcels.gml', None)]
```
I think these councils can be matched after a little edit in `_standardise_list_council_names` (which I can't comment on because its not in these PR changes). I changed it to:

```
def _standardise_list_council_names(name_series: pd.Series) -> list:
    """
    Standardise council (local authority district) names.

    Args:
        name_series (pd.Series): council names

    Returns:
        list: standardised council names
    """
    council_names = [
        # Replace spaces, full stops, numbers, other punctuation with "_"
        re.sub("[^a-zA-Z-,]+", "_", nm).lower().split(",")[0]
        for nm in name_series
    ]

    manual_rename = {
        'county_durham': 'durham_county',
        'king_s_lynn_and_west_norfolk': 'kings_lynn_and_west_norfolk',
        'kingston_upon_hull': 'hull_city',
        'newcastle_upon_tyne': 'newcastle_city'
    }

    council_names = [manual_rename.get(c, c) for c in council_names]

    return council_names
```

File: asf_heat_pump_suitability/pipeline/prepare_features/land_extent.py [hunk 9]
```diff
@@ -50,57 +78,52 @@ def _standardise_list_council_names(name_series: pd.Series) -> list:
...
         council/LAD name, council/LAD code, council/LAD geometry.
-        save_as (str): path to save matched files to. Optional.
 
     Returns:
         gpd.GeoDataFrame: land extent (INSPIRE) files and their bounding polygons in British National Grid CRS
     """
-    council_bounds = transform_gdf_council_bounds(ladnm_col, use_cols)
-    land_extent_file_names = base_getters.list_files_s3_location(land_extent_location)
+    bounds_gdf = transform_gdf_council_bounds(ladnm_col, use_cols)
+    files = base_getters.list_obj_s3_location(path)
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/prepare_features/land_extent.py:114]: When I did this, one of the "files" was the path directory itself i.e. `'asf-heat-pump-suitability/source_data/inspire_gml/'`, so maybe:
```suggestion
    files = base_getters.list_obj_s3_location(path)
    files = [file for file in files if file.endswith(".gml")]
```

**Review by lizgzil — COMMENTED** [review]
Thanks for all this @crispy-wonton 🏴󠁧󠁢󠁳󠁣󠁴󠁿 

I'm splitting up my reviews into one batch for each of the features.

This set is for the garden size estimates - the code looks good! I can't see anything that would have incorrectly measured the garden size at big, but have a suggestion that means that `fill_nulls_file_bounds` might not need to be run

File: asf_heat_pump_suitability/pipeline/prepare_features/property_density.py [hunk 10]
```diff
@@ -61,3 +61,55 @@ def replace_zeros_with_none_df(df: pl.DataFrame, column_name: str) -> pl.DataFra
...
+    Generate dataframe with property density (dwellings per km2) per 2011 Data Zone in Scotland. Long-term empty dwellings
+    are excluded from dwelling count to calculate property density.
+
+    Returns:
+        pl.DataFrame: properties per km2 per 2011 Data Zone in Scotland
+    """
+    dz_df = load_transform_df_datazone_area()
+    dwellings_df = load_transform_df_n_dwellings_s()
+    df = dwellings_df.join(dz_df, how="inner", on="DataZone").with_columns(
+        (pl.col("n_dwellings") / pl.col("StdAreaKm2")).alias("property_density_km2")
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/prepare_features/property_density.py]: For england and wales this was called "Property density (households per KM2)", so I assume they need to be consistent?

File: asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py (hunk 6 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py]: Yes that's the reason. It's because we label the source data dirs by EPC year and quarter. I thought it would be easier to label the sources this way so that we know which ones to use with which EPC version.

And you're right. Otherwise, this has nothing to do with EPC files.

File: asf_heat_pump_suitability/pipeline/prepare_features/property_density.py (hunk 10 above)
Comment (lizgzil) [reply asf_heat_pump_suitability/pipeline/prepare_features/property_density.py]: I had a look into the 'StdAreaKm2' column, in reading this:
https://geoportal.statistics.gov.uk/documents/f6717e32fb9c4304988c2c30bf063b1b/explore
it says "The recommended approach preferred by Eurostat is to use the ‘Land Area’ measurement
(Area to Mean High Water Excluding Area of Inland Water) to compile population density
measurements."

I see in the england and wales method the land area column is called "Land Count (Area in KM2)" - which doesn't suggest it's standardised? So not sure if there is an inconsistency here. @danlewis85 what do you think?

Reassuringly, I compared the results and they aren't much different:
Screenshot 2024-11-08 at 17 05 22

Stats when grouped by lsoa and population density averaged per lsoa:

Screenshot 2024-11-08 at 17 27 40

File: asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py [hunk 11]
```diff
@@ -139,28 +144,23 @@ def parse_arguments() -> argparse.Namespace:
     logging.info("Adding property density to EPC")
     epc_df = property_density.extend_df_with_property_density(epc_df)
 
+    # Add feature: property density Scotland
+    dz_density_df = property_density.generate_df_property_density_s()
+    epc_df = epc_df.join(dz_density_df, how="left", left_on="lsoa", right_on="DataZone")
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:149]: I was looking at the data in "s3://asf-heat-pump-suitability/outputs/2023Q4/20241105_2023_Q4_EPC_weighted_features.parquet" and there is no 'DataZone' column. I guess at some point you removed it (I couldn't see in this code why it wouldn't be saved out)? I don't think it matters, but thought I'd mention it in case it indicates an unintended issue.

File: asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py (hunk 7 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py:67]: We do have an existing version of the file bounds for England and Wales, it just has a different name at the moment: `2023_land_parcels_with_file_polygons.geojson`. I need to run this script for England and Wales so that it produces outputs with the expected names. I just hadn't done it yet because the `_fill_null_file_bounds` function takes a long time to run and I was focussing on Scotland.

Currently, `run_scripts/run_calculate_garden_size.py` will fail if running for England and Wales because it will look for a mapping file with a different name. This can be updated in the script. I will run this `run_get_inspire_file_bounds.py` for England and Wales now though to generate the appropriate file and for testing purposes.

File: asf_heat_pump_suitability/pipeline/prepare_features/land_extent.py (hunk 9 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/prepare_features/land_extent.py:114]: Yes this is weird. I had previously written the code to cope with this case, but when I applied the same code to the Scotland directory, it didn't have the behaviour of including the dir name so I thought it was a previous bug that had somehow disappeared. Now i see it's back. Not sure why they have different behaviours with the same function.

The Scotland dir is full of subdirs but the England and Wales dir is full of files which is the only thing I can imagine might make a difference.

Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/prepare_features/land_extent.py:114]: I've committed your suggestion.

**Review by lizgzil — COMMENTED** [review]
Thanks @crispy-wonton this is great and so much work 🎉 

I think I've looked through everything and had a few clarification questions doted around the place.

Going through your review Qs:

#### `check appropriate datasets used and that they have been processed correctly` 
They look good as far as I can see, potentially some changes needed to column names to make them consistent across E/W and S data. I did a bit of analysis comparing the `2023Q4/20241105_2023_Q4_EPC_weighted_features.parquet` E/W properties vs S properties and they look ok. Obviously a few differences, but nothing shocking.

#### `garden size estimate pipeline for Scotland to ensure it's been calculated correctly as the results are pretty large average garden sizes (median is over 1000m2).`
See my first review.

#### `appropriate dwelling total has been used to calculate property density. Should we exclude any other dwelling categories from the total count?`

I downloaded the data and to make the numbers add up it looks like `"Total number of dwellings" = "Occupied dwellings" + "Unoccupied dwellings exempt from paying Council Tax" + "Long-term empty dwellings 
[Note 6]" + "Second homes"` ("Vacant dwellings" aren't included)
(perhaps that was already known!)

Since Dan said that "Second homes" should be included, that means it's just "Unoccupied dwellings exempt from paying Council Tax" which is questionable of whether to remove or not. And I think since these seem to be <3% its not worth worrying about.

#### `Double check all required features have been added`

I think so! 👍

File: asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py (hunk 7 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/run_scripts/run_get_inspire_file_bounds.py:67]: The `inspire_file_bounds_ew.geojson` is now available in s3.

File: asf_heat_pump_suitability/pipeline/prepare_features/property_density.py (hunk 10 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/prepare_features/property_density.py]: Good catch on the column names. I have updated them both to `"households_per_km2"` as I think it's most accurate while remaining short.
I will look into the datasets further to see if it's already standardised or if a standardised one is available.

Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/prepare_features/property_density.py]: I have checked and I believe the columns used (`StdAreaKm2` for Scotland and `Land Count (Area in KM2)` for England and Wales) represent the standard area measurements of "area to mean high water excluding area of inland water".

Scotland:
I compared the Scottish data to this land area dataset in hectares from the Scottish government (description: "The **land area to mean high water excluding areas of inland water**, in hectares. It is based on aggregating 2011 Data Zone data. The Data Zone data was created by aggregating National Records of Scotland 2011 Census Output Areas."' source: [https://statistics.gov.scot/resource?uri=http%3A%2F%2Fstatistics.gov.scot%2Fdata%2Fland-area-2011-data-zone-based]). After converting Ha to Km2, the measurements in this dataset compared to the one used in the pipeline [source: https://www.data.gov.uk/dataset/ab9f1f20-3b7f-4efa-9bd2-239acf63b540/data-zone-boundaries-2011], the area measurements are the same except for differences in rounding, with the one used in the pipeline having more decimal places.

England and Wales:
For Scotland and Wales, we use the `Land count (Area in Km2)` measurement. The other land area columns in the dataset include the following: `Extent of the Realm (Area in KM2)`; `Clipped to the Coastline (Area in KM2)`; and `Area of Inland Water (KM2)`. These are different from the columns mentioned in the documentation. [Source: https://geoportal.statistics.gov.uk/datasets/a488cb8fc9a74accb63cb52961e456ef/about]

The documentation states the following:
"Four types of measurements are included: total extent (AREAEHECT), area to mean high water (coastline) (AREACHECT), area of inland water (AREAIHECT) and area to mean high water excluding area of inland water (land area) (AREALHECT). **The Eurostat-recommended approach is to use the ‘land area’ measurement to compile population density figures.**"

[not sure what happened to text formatting]`AREALHECT` is described as 'area to mean high water excluding area of inland water'. Since `AREACHECT` is labelled as 'area to mean high water' and `AREAIHECT` is labelled as 'area of inland water', this implies `AREALHECT` (the column of interest) is equal to `AREACHECT - AREAIHECT`. In the dataset, if we calculate: 
`Clipped to the Coastline (Area in KM2) - Area of Inland Water (KM2)` then we get values equal to those in the `Land Count (Area in KM2)` column.
--------
In conclusion, I think this confirms that both data sources we are using represent the required standard area measurements for calculating population density.

File: asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py (hunk 11 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:149]: I think this is normal & expected behaviour. 'DataZone' is what we are using as the Scottish equivalent to 'LSOA'. Scottish DataZone codes appear in the `lsoa` column in EPC. When we do the join here, that's why we are joining on `lsoa` col from EPC and `DataZone` col from Scottish property density df. I think it's normal that the column to join on does not get duplicated in the resulting dataframe? So I don't think this is a problem but let me know if you still have concerns

File: asf_heat_pump_suitability/pipeline/suitability/calculate_suitability.py (hunk 4 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/suitability/calculate_suitability.py:200]: **CORRECTION**:
This conversion is no longer needed because `pipeline/run_scripts/run_add_features.py` has been been updated to output a flag with `True`/`False` already for listed building status.

File: asf_heat_pump_suitability/pipeline/prepare_features/land_extent.py (hunk 8 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/prepare_features/land_extent.py:117]: Thank you , this is a great suggestion which I have added. I kept the `fill_null_file_bounds` function in the function to generate file bounds because I think the INSPIRE files are missing a few local authorities at the source, so those can be filled if needed if they ever get added and we re-run this pipeline.

I also added a new assertion statement to ensure that the pipeline errors out if there are more than 10% of .gml files with no file bounds.

**Comment by crispy-wonton** [comment]
Thanks @lizgzil for your review. In response to your comment:

> I downloaded the data and to make the numbers add up it looks like "Total number of dwellings" = "Occupied dwellings" + "Unoccupied dwellings exempt from paying Council Tax" + "Long-term empty dwellings  [Note 6]" + "Second homes" ("Vacant dwellings" aren't included)
> (perhaps that was already known!)
> 
> Since Dan said that "Second homes" should be included, that means it's just "Unoccupied dwellings exempt from paying Council Tax" which is questionable of whether to remove or not. And I think since these seem to be <3% its not worth worrying about.

We agreed to exclude 'long-term empty dwellings' and include 'second homes' as you've said. I've only excluded 'long-term empty' and no others.

File: asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py (hunk 11 above)
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:149]: I have two columns - `households_per_km2` and `households_per_km2_right` when I ran this script, I think this join is where it happened. I assume they'll need to be merged into one?

File: asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py [hunk 12]
```diff
@@ -139,28 +144,23 @@ def parse_arguments() -> argparse.Namespace:
...
-    logging.info("Loading listed buildings for Wales")
-    w_listed_buildings_df = listed_buildings.transform_gdf_listed_buildings("Wales")
-    w_listed_buildings_df = listed_buildings.sjoin_df_epc_with_listed_buildings(
-        epc_df, w_listed_buildings_df
-    )
-
-    listed_buildings_df = pl.concat(
-        [e_listed_buildings_df, w_listed_buildings_df], how="vertical"
+    epc_df = epc_df.join(listed_buildings_df, how="left", on="UPRN").with_columns(
+        pl.col("listed_building").fill_null(False)
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:165]: I'm not sure where it happened but I also have two columns `UPRN` and `UPRN_right`. It seems that all the english data has null for `UPRN`.

**Review by lizgzil — COMMENTED** [review]
Thanks @crispy-wonton ! I ran `run_add_features.py` and it looks good and ran, but there were two columns with `_right` suffixes which probably should be fixed

**Review by lizgzil — DISMISSED** [review]
`python -i asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_garden_size.py --epc_path s3://asf-heat-pump-suitability/outputs/2023Q4/20240827_2023_Q4_EPC_weighted_features.parquet -y 2023 -q 4 -n ew ` ran for 8 iterations without error, so I think it's good to go! :)

File: asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py (hunk 11 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:149]: Great spot thank you!

File: asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py (hunk 12 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:165]: TLDR: this will be fixed in a separate PR because it originates in the reweighting script not the add features script.

Thanks for noticing this! I looked into it. It seems to come from the reweighting pipeline because it already exists in the dataset that we load in at the beginning of `run_add_features.py`. I think it originates from the [outer join](https://github.com/nestauk/asf_heat_pump_suitability/blob/a75c867fe0b4ef85182e4d90292a3cb1edeb1693/asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py#L151) we use to join the weights to the `epc_df`. This is an expected `polars` behaviour that I forgot to account for. 

This means, rows which have nulls in `UPRN_right` should be missing weights. They will either be missing weights because:

- They are in an LSOA which was skipped because we are missing census data for it (mainly affects Scotland, but also England and Wales in some instances)
- We had to drop the row for weighting because the row had a category which was not found in the target data for that LSOA. E.g. if the property type of the row is 'flat' but the census has 0% flats for that LSOA, the row will be dropped.
- They are in Scotland. This dataset was run before reweighting for Scotland was added, so all Scottish rows should be missing a weight.

Here is the count of rows with missing `UPRN_right` for each country:
```
>>> epc.filter(pl.col("UPRN_right").is_null())["COUNTRY"].value_counts()
shape: (3, 2)
┌──────────┬─────────┐
│ COUNTRY  ┆ count   │
│ ---      ┆ ---     │
│ cat      ┆ u32     │
╞══════════╪═════════╡
│ Wales    ┆ 38285   │
│ England  ┆ 788475  │
│ Scotland ┆ 1473612 │
```

All of them have null weights, as expected:
```
>>> epc.filter(pl.col("UPRN_right").is_null())["weight"].value_counts()
shape: (1, 2)
┌────────┬─────────┐
│ weight ┆ count   │
│ ---    ┆ ---     │
│ f64    ┆ u32     │
╞════════╪═════════╡
│ null   ┆ 2300372 │
```
